### PR TITLE
HPCC-14873 Add support for parallel aggregates

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday gavin.halliday@lexisnexis.com

### CODE REVIEW

**Comment by ghalliday** [comment]
@jakesmith 
@richardkchapman 
This is now close to being ready to pull.  The regression suite now passes forcing all stranded activities to their parallel forms.  I think there may be some issues with counts not appearing on the graph which I will investigate, but otherwise it is ready for an initial code review.

File: common/thorhelper/thorstrand.cpp [hunk 1]
```diff
@@ -332,6 +393,9 @@ class StreamFromBlockQueue : public CInterfaceOf<IEngineRowStream>
 
     virtual void stop()
     {
+        //A stop on any unordered branch implies that there is no need to read any more on other threads?
+        //??Is this strictly correct?
```
Comment (richardkchapman) [inline common/thorhelper/thorstrand.cpp]: Worth finding out the answer before I merge?

**Comment by richardkchapman** [comment]
@ghalliday Back in your court atm

**Comment by ghalliday** [comment]
@richardkchapman please can you take a careful look at these changes.  Especially the implementation of the new functions for possible race conditions.

**Comment by ghalliday** [comment]
It looks like the regression suite passed with it enabled.

File: system/jlib/jqueue.hpp [hunk 2]
```diff
@@ -367,12 +387,15 @@ class ReaderWriterQueue
...
     inline bool allWritersStopped() const { return activeWriters.load(std::memory_order_acquire) <= 0; }
+    inline bool allReadersStopped() const { return activeReaders.load(std::memory_order_acquire) <= 0; }
 
 protected:
     BufferElement * values;
     unsigned dynamicSlotMask;
     unsigned maxItems;
+    unsigned maxReaders;
     unsigned maxWriters;
+    std::atomic<int> activeReaders;
```
Comment (richardkchapman) [inline system/jlib/jqueue.hpp]: Why is maxReaders unsigned but activeReaders int?

Comment (ghalliday) [inline system/jlib/jqueue.hpp]: I wanted to safely recover from underflow.  Hopefully it will never occur.

**Pull request merged**